pandas/core/index.py

```diff
@@ -2510,7 +2527,342 @@ def invalid_op(self, other=None):
+                data = data.set_categories(categories)
+            if not data.ordered:
+                data = data.as_ordered()
+            return data
+
+        if fastpath:
+            return cls._simple_new(data, name=name)
+
+        if ordered is not None:
+            raise ValueError("CategoricalIndex are by definition ordered")
+
+        if isinstance(data, ABCCategorical):
+            data = create_categorical(data, categories)
+        elif isinstance(data, CategoricalIndex):
+            data = data._data
+            data = create_categorical(data, categories)
+        else:
+
+            # don't allow scalars
+            # if data is None, then categories must be provided
+            if lib.isscalar(data):
+                if data is not None or categories is None:
+                    cls._scalar_data_error(data)
+                data = []
+            from pandas.core.categorical import Categorical
+            data = Categorical(data, categories=categories, ordered=True)
+
+        if copy:
+            data = data.copy()
+
+        return cls._simple_new(data, name=name)
+
+    @classmethod
+    def _simple_new(cls, values, name=None, categories=None, **kwargs):
+        result = object.__new__(cls)
+
+        if not isinstance(values, ABCCategorical):
+            from pandas.core.categorical import Categorical
+            values = Categorical(values, categories=categories, ordered=True)
+        elif categories is not None:
+            values = values.set_categories(categories)
+
+        result._data = values
+        result.name = name
+        for k, v in compat.iteritems(kwargs):
+            setattr(result,k,v)
+
+        result._reset_identity()
+        return result
+
+    def equals(self, other):
+        """
+        Determines if two CategorialIndex objects contain the same elements.
+        """
+        if self.is_(other):
+            return True
+
+        try:
+            if is_categorical_dtype(other):
+                if not self.categories.equals(other.categories):
+                    return False
+            else:
+                from pandas import Categorical
+                other = Categorical(other, categories=self.categories, ordered=True)
+                if isnull(other).any():
+                    return False
+
+            return (self._data == other).all()
+        except:
+            pass
+
+        return False
+
+    @property
+    def inferred_type(self):
+        return 'categorical'
+
+    @property
+    def values(self):
+        """ return the underlying data, which is a Categorical """
+        return self._data
+
+    @property
+    def codes(self):
+        return self._data.codes
+
+    @property
+    def categories(self):
+        return self._data.categories
+
+    @property
+    def ordered(self):
+        return self._data.ordered
+
+    def __contains__(self, key):
+        hash(key)
+        return key in self.categories
+
+    def __array__(self, result=None):
```

Comment:
hmm, ok, we don't do this for Series, i'll fix for Categorical/CategoricalIndex here, maybe make a separate issue for this